### PR TITLE
minestom-ce pull-back

### PR DESCRIPTION
**Pull request by mworzala**
This PR brings back most of the changes from [minestom-ce](https://github.com/hollow-cube/minestom-ce), though some cases the changes are modified.

### Changes (at a high level):
* 1.19.4 - 1.20.4
* Gradle rework
  * Simplification
  * Maven central publishing (needs some work still, see below)
* BlockPlacementRule update
  * Support most vanilla use cases
  * Remove existing rules (will probably either publish the set I have separately or add them here)
* Extensions Removed (You can have the same behavior using the [extension addon](https://github.com/hollow-cube/minestom-ce-extensions))
* [And more...](https://github.com/hollow-cube/minestom-ce/blob/main/CHANGELOG.md#changelog)

### Maven Central
How to handle maven releases is a bit up in the air, in my mind there are two competing options described below. Please provide feedback/opinions in this thread or in discord if you have any.

#### Split snapshots/releases

Snapshots (either located at `net.minestom:minestom-snapshots:{version}`, or at `net.minestom:minestom:{version}` in the sonatype staging repository) will be both commit tagged (eg `d11cee55d2`) and PR-branch tagged (eg `my_pr_branch-d11cee55d2`).

Releases (located at `net.minestom:minestom:{version}`) will be tagged to specific game versions, with either a date based versioning scheme (eg `2023.9.4+1.20.1`) or a semantic version (eg `1.2.3+1.20.1`). These versions would presumedly be significantly less frequent and include more changes. This will likely come with a later PR.

#### Single release

Releases (at `net.minestom:minestom:{version}`) will be added when the repository is tagged manually with a new version only. They would follow a semantic version tied to a game version, eg `1.2.3+1.20.1`. Potentially PR branches would be published to another repo as well.

Closes #1718

### CODE REVIEW

**Comment by Nuckerr** [comment]
Imo, do split snapshots/releases. Without that, versions would be fractured over multiple repositories, better to keep it on the same repo.

**Comment by mworzala** [comment]
What do you mean they would be fractured over repositories?

**Comment by Nuckerr** [comment]
You would have to go to jitpack for snapshots, and central for releases (unless I've misunderstood)

**Comment by mworzala** [comment]
In any case jitpack is going away

File: src/main/java/net/minestom/server/world/biomes/Biome.java [hunk 1]
```diff
@@ -72,6 +72,7 @@ public static Builder builder() {
                 element.setFloat("scale", scale);
                 element.setFloat("downfall", downfall);
                 element.setString("category", category.name().toLowerCase(Locale.ROOT));
+                element.setByte("has_precipitation", (byte) 1); //todo bad fix
```
Comment (GreatWyrm) [inline src/main/java/net/minestom/server/world/biomes/Biome.java:75]: Could address this quickly with another member variable in this class, unless you have other things in mind to do.

**Review by GreatWyrm — CHANGES_REQUESTED** [review]
Alright went through a bunch of the files and left my comments.

Overall, I like the changes, although I have a few small concerns over it.

File: build.gradle.kts [hunk 2]
```diff
@@ -37,63 +93,92 @@ tasks {
...
-    api(libs.hydrazine)
+        repositories.sonatype {
+            nexusUrl.set(uri("https://s01.oss.sonatype.org/service/local/"))
+            snapshotRepositoryUrl.set(uri("https://s01.oss.sonatype.org/content/repositories/snapshots/"))
 
-    // Adventure, for user-interface
-    api(libs.bundles.adventure)
+            if (System.getenv("SONATYPE_USERNAME") != null) {
+                username.set(System.getenv("SONATYPE_USERNAME"))
+                password.set(System.getenv("SONATYPE_PASSWORD"))
```
[the remark this answers is not in the document]
Comment (mworzala) [reply build.gradle.kts:121]: yes

File: demo/src/main/java/net/minestom/demo/Main.java [hunk 3]
```diff
@@ -23,11 +24,13 @@
...
     public static void main(String[] args) {
+        System.setProperty("minestom.use-new-chunk-sending", "true");
+
         MinecraftServer minecraftServer = MinecraftServer.init();
 
         BlockManager blockManager = MinecraftServer.getBlockManager();
-
-        blockManager.registerBlockPlacementRule(new RedstonePlacementRule());
+        blockManager.registerBlockPlacementRule(new DripstonePlacementRule());
+        blockManager.registerHandler(TestBlockHandler.INSTANCE.getNamespaceId(), () -> TestBlockHandler.INSTANCE);
```
[the remark this answers is not in the document]
Comment (mworzala) [reply demo/src/main/java/net/minestom/demo/Main.java:33]: meh its the demo, half of it is test code

File: src/main/java/net/minestom/server/ServerFlag.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    // Online Mode
+    public static final @NotNull String AUTH_URL = System.getProperty("minestom.auth.url", "https://sessionserver.mojang.com/session/minecraft/hasJoined");
+
+    // World
+    public static final @Nullable String STACKING_RULE = System.getProperty("minestom.stacking-rule");
+    public static final int WORLD_BORDER_SIZE = Integer.getInteger("minestom.world-border-size", 29999984);
+
+    // Terminal
+    public static final boolean TERMINAL_ENABLED = System.getProperty("minestom.terminal.disabled") == null;
+    public static final boolean TERMINAL_SUPPORT_HEX_COLOR = PropertyUtils.getBoolean("minestom.terminal.support-hex-color", true);
```
[the remark this answers is not in the document]
Comment (mworzala) [reply src/main/java/net/minestom/server/ServerFlag.java:44]: Boolean.getBoolean does not support a default

File: src/main/java/net/minestom/server/command/builder/arguments/Argument.java [hunk 5]
```diff
@@ -202,6 +204,12 @@ public Supplier<T> getDefaultValue() {
      */
     @NotNull
     public Argument<T> setDefaultValue(@Nullable Supplier<T> defaultValue) {
+        this.defaultValue = unused -> defaultValue.get();
```
[the remark this answers is not in the document]
Comment (mworzala) [reply src/main/java/net/minestom/server/command/builder/arguments/Argument.java:207]: this.defaultValue is a function not a supplier

**Comment by mworzala** [comment]
For now I will set it to publish commits to `minestom-snapshots`, it can be changed in the future.

**Comment by MelonHell** [comment]
What's up with this pull request now?

**Comment by mworzala** [comment]
Still planned, just haven't gotten to updating it yet.

**Comment by DasBabyPixel** [comment]
Terminal and extension removal getting merged?

**Comment by DasBabyPixel** [comment]
Publishing gonna get fixed so we can get Minestom on central?

**Comment by mworzala** [comment]
Yes.

**Pull request merged**